### PR TITLE
feat: Readable metrics view url params

### PR DESCRIPTION
**Pull request by AdityaHegde**
Move our dashboard url state from base64 encoded protobuf to human readable urls params.

- [x] Updated V1ExplorePreset to include all fields.
- [x] Replace protobuf with a url params translator.
- [x] Handle non-standard filters in all network calls.
- [x] Show a free-form editor for non-standard filters.
- [x] Show a readonly pill when non-standard filter is selected.
- [x] Add all fields to url state.
- [x] Do not add pivot fields when it is not active.
- [x] Fix unit tests
- [x] Fix e2e tests
- [x] Add docs (https://www.notion.so/rilldata/Human-readable-state-docs-for-SewerAI-145ba33c8f57805f8d15c85141e545ac)
- [ ] Figure out expected behaviour around visual explore editor. (Follow up PR)
- [ ] Add more exhaustive tests. (Follow up PR)

### CODE REVIEW

File: web-common/src/features/dashboards/url-state/getBasePreset.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  V1ExploreComparisonMode,
+  type V1ExplorePreset,
+  type V1ExploreSpec,
+  V1ExploreWebView,
+} from "@rilldata/web-common/runtime-client";
+
+export function getBasePreset(
+  explore: V1ExploreSpec,
+  preferences: LocalUserPreferences,
+) {
```
Comment (ericpgreen2) [inline web-common/src/features/dashboards/url-state/getBasePreset.ts:19]: It feels like this function is doing two things: 1. defining the "base" ("default"?) state for all Explores and 2. merging the generic default state with the given Explore's default state and the user preferences.

Could we define a `DEFAULT_EXPLORE_STATE` in a constant outside this function? And maybe the function name should be reconsidered.

File: web-common/src/features/dashboards/url-state/url-state.spec.ts [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+        const basePreset = getBasePreset(explore, {
+          timeZone: "UTC",
+        });
+        const initEntity = getDefaultMetricsExplorerEntity(
+          AD_BIDS_EXPLORE_NAME,
+          AD_BIDS_METRICS_3_MEASURES_DIMENSIONS,
+          explore,
+          AD_BIDS_TIME_RANGE_SUMMARY,
+        );
+        cleanMetricsExplore(initEntity);
```
Comment (ericpgreen2) [inline web-common/src/features/dashboards/url-state/url-state.spec.ts:389]: Why is this necessary?

File: web-common/src/features/dashboards/url-state/url-state.spec.ts [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+
+        // assert that the entity we got back matches the original
+        expect(entityFromDefaultUrl).toEqual(initEntity);
+      });
+    }
+  });
+});
+
+// cleans up any UI only state from MetricsExplorerEntity
+export function cleanMetricsExplore(
```
Comment (ericpgreen2) [inline web-common/src/features/dashboards/url-state/url-state.spec.ts:490]: This comment implies the function name could be `removeUIStateFromMetricsExplorerEntity`. But doesn't the `MetricsExplorerEntity` contain _all_ UI/client-side state?

File: web-common/src/features/dashboards/user-preferences.ts [hunk 4]
```diff
@@ -40,6 +40,25 @@ function localUserPreferencesActions() {
...
+    try {
+      return JSON.parse(localstorageUserPreference);
+    } catch {
+      // TODO
+    }
+  }
+
+  return {
+    timeZone: getLocalIANA(),
+  };
```
Comment (ericpgreen2) [inline web-common/src/features/dashboards/user-preferences.ts:59]: I wouldn't expect this default/fallback to be included in this function. I'd expect `timeZone: getLocalIANA()` to be included in the default generic Explore state, which would later be merged with user preferences.

File: web-local/src/routes/(viz)/explore/[name]/+layout.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  try {
+    const { explore, metricsView, basePreset } = await fetchExploreSpec(
+      instanceId,
+      exploreName,
+    );
+
+    return {
+      explore,
+      metricsView,
+      basePreset,
```
Comment (ericpgreen2) [inline web-local/src/routes/(viz)/explore/[name]/+layout.ts:22]: What about `initialExploreState`?

File: web-local/src/routes/(viz)/explore/[name]/+page.svelte [hunk 6]
```diff
@@ -16,20 +15,21 @@
   const queryClient = useQueryClient();
 
   export let data: PageData;
+  $: ({ metricsView, explore, basePreset, partialMetrics } = data);
```
Comment (ericpgreen2) [inline web-local/src/routes/(viz)/explore/[name]/+page.svelte:18]: It's confusing seeing both `basePreset` and `partialMetrics`. Could we merge the two of them further upstream?

File: web-admin/src/routes/[organization]/[project]/-/reports/[report]/open/+page.svelte
[the remark this answers is not in the document]
Comment (AdityaHegde) [reply web-admin/src/routes/[organization]/[project]/-/reports/[report]/open/+page.svelte:1]: Yes it will just add `state` param for older reports that relied on it.

File: web-common/src/features/dashboards/url-state/url-state.spec.ts (hunk 2 above)
Comment (AdityaHegde) [reply web-common/src/features/dashboards/url-state/url-state.spec.ts:389]: Not all data in MetricsExploreEntity is stored into URL. This cleans it up to simplify comparison. I will rename it to be clearer.

File: web-local/src/routes/(viz)/explore/[name]/+page.svelte (hunk 6 above)
Comment (AdityaHegde) [reply web-local/src/routes/(viz)/explore/[name]/+page.svelte]: They are not used together. If `basePreset` is used to keep the url clean and not add defaults. `partialMetrics` is the metrics from url. `basePreset` will not be needed in the future when we directly update the url in actions.

**Review by ericpgreen2 — COMMENTED** [review]
I kicked off a [UXQA doc here](https://www.notion.so/rilldata/Human-readable-state-UXQA-146ba33c8f5780feb362ca46695c6179?d=146ba33c8f578069bd09001c6e2f10fc#146ba33c8f578016913ae9b2fd62d723)

**Review by ericpgreen2 — CHANGES_REQUESTED** [review]
- As discussed, there should be one place where we can find all available URL query parameters
- Can you mark `dashboard.proto` as deprecated and include a note that we previously used it to serialize state for URLs?

**Review by ericpgreen2 — COMMENTED** [review]
One reason our state management is complex is because we have:
a) Many sources of state (which is unavoidable, per our requirements)
b) State merging happening in multiple, disconnected places. We have a waterfall of state merges.

To simplify, I’d propose we **merge the state together in one canonical place**, instead of the current waterfall approach.

I count these sources of state (in increasing order of precedence): 
- Rill-opinionated default
- developer-specified preset
- user time zone preference
- home bookmark
- magic token state
- local storage ("last viewed state")
- session storage
- URL query parameters

Some discrete issues:
  - `getDefaultExplorePreset()` does a subset of state merge. It merges the Rill-opinionated default with the time zone user preference with the developer-specified default.
  - `convertURLToExplorePreset()` does another subset of state merge. It merges URL parameters with the `defaultExplorePreset`.
  - `convertSourceXToExplorePreset()` shouldn’t be necessary. We shouldn’t have to convert anything into an `ExplorePreset`. We should only have to convert state sources into `Partial<ExploreState>`.

Suggestions:
- Replace `convertSourceXToSourceY()` functions with `getPartialStateFromSourceX()` functions
- Add one canonical merge: `const state = mergeState(sourceXPartialState, sourceYPartialState, ...)`
- List sources in priority order from lowest (Rill defaults) to highest (URL params)

File: web-admin/src/features/bookmarks/BookmarksDropdownMenuContent.svelte [hunk 7]
```diff
@@ -39,7 +57,14 @@
     metricsViewName,
     exploreName,
   );
-  $: filteredBookmarks = searchBookmarks($bookmarks.data, searchText);
+  $: filledInBookmarks = getFilledInBookmarks(
+    $bookmarks.data,
+    `/${organization}/${project}/explore/${exploreName}`,
```
[the remark this answers is not in the document]
Comment (AdityaHegde) [reply web-admin/src/features/bookmarks/BookmarksDropdownMenuContent.svelte]: Updated

File: web-admin/src/features/bookmarks/selectors.ts [hunk 8]
```diff
@@ -25,8 +32,10 @@ import { derived, type Readable } from "svelte/store";
 
 export type BookmarkEntry = {
```
[the remark this answers is not in the document]
Comment (AdityaHegde) [reply web-admin/src/features/bookmarks/selectors.ts:27]: Ya just the url is needed

File: web-common/src/features/dashboards/stores/metrics-explorer-entity.ts [hunk 9]
```diff
@@ -74,7 +74,6 @@ export interface MetricsExplorerEntity {
   sortDirection: SortDirection;
 
   whereFilter: V1Expression;
-  havingFilter: V1Expression;
```
[the remark this answers is not in the document]
Comment (AdityaHegde) [reply web-common/src/features/dashboards/stores/metrics-explorer-entity.ts:77]: We never used it. Url with `having` is converted to `dimensionThresholdFilters`

File: web-common/src/lib/url-utils.ts [hunk 10]
```diff
@@ -24,3 +24,12 @@ export function getFullUrlForPath(
...
 }
+
+export function mergeSearchParams(
+  fromSearchParams: URLSearchParams,
+  toSearchParams: URLSearchParams,
+) {
+  fromSearchParams.forEach((value, key) => {
+    toSearchParams.set(key, value);
+  });
+}
```
[the remark this answers is not in the document]
Comment (AdityaHegde) [reply web-common/src/lib/url-utils.ts:35]: This is just a utility function to merge 2 search params. Returning a new one is redundant.

File: web-common/src/features/dashboards/url-state/ExploreWebViewStore.ts [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+        ExploreWebViewNonPivot,
+        exploreState,
+        exploreSpec,
+        view,
+      );
+    }
+    this.updateStoreForView(view, exploreState, exploreSpec);
+  }
+
+  public getUrlForView(
```
[the remark this answers is not in the document]
Comment (AdityaHegde) [reply web-common/src/features/dashboards/url-state/ExploreWebViewStore.ts]: Removed

File: web-local/src/routes/(viz)/explore/[name]/+layout.ts [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+import { runtime } from "@rilldata/web-common/runtime-client/runtime-store";
+import { error } from "@sveltejs/kit";
+import { get } from "svelte/store";
+
+export const load = async ({ params, depends }) => {
+  const { instanceId } = get(runtime);
+
+  const exploreName = params.name;
+
+  depends(exploreName, "explore");
```
[the remark this answers is not in the document]
Comment (AdityaHegde) [reply web-local/src/routes/(viz)/explore/[name]/+layout.ts:11]: Its in a lot of places so lets do that separately.

**Review by ericpgreen2 — APPROVED** [review]
